oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMergeSortTaskTest.java

```diff
@@ -19,103 +19,134 @@
-        assertEquals("The sort work directory should contain only the flat file store, the intermediate files should have been deleted after merged. Instead it contains: " + Arrays.toString(filesInWorkDir),
-                1, filesInWorkDir.length);
-        assertTrue(result.getFlatFileStoreFile().exists());
+
+        try (Stream<Path> fileStream = Files.list(sortRoot)) {
+            List<String> filesInWorkDir = fileStream
+                    .map(path -> path.getFileName().toString())
+                    .collect(Collectors.toList());
+            assertEquals("The sort work directory should contain only the flat file store, the intermediate files should have been deleted after merged. Instead it contains: " + filesInWorkDir,
+                    1, filesInWorkDir.size());
+        }
+        assertTrue(Files.exists(result.getFlatFileStoreFile()));
         return result;
     }
 
     @Test(expected = IllegalStateException.class)
     public void badInputFile() throws Exception {
-        File singleFileToMerge = createFileWithWrongFormat();
-        runTest(singleFileToMerge);
+        Path singleFileToMerge = createFileWithWrongFormat();
+        runTest(algorithm, singleFileToMerge);
     }
 
-    private File createFileWithWrongFormat() throws Exception {
-        File file = Files.createTempFile("merge-stage-input", ".json").toFile();
-        try (BufferedWriter bw = Files.newBufferedWriter(file.toPath(), FLATFILESTORE_CHARSET)) {
+    private Path createFileWithWrongFormat() throws Exception {
+        Path file = Files.createTempFile(sortFolder.getRoot().toPath(), "merge-stage-input", ".json");
+        try (BufferedWriter bw = Files.newBufferedWriter(file, FLATFILESTORE_CHARSET)) {
             bw.write("/a/b/c\n");
         }
-        file.deleteOnExit();
         return file;
     }
+
+    @Test
+    public void manyFilesToMergeDidNotMerge() throws Exception {
+        int intermediateFilesCount = 256;
+        System.setProperty(PipelinedMergeSortTask.OAK_INDEXER_PIPELINED_EAGER_MERGE_TRIGGER_THRESHOLD, "20");
+        System.setProperty(PipelinedMergeSortTask.OAK_INDEXER_PIPELINED_EAGER_MERGE_MAX_FILES_TO_MERGE, "1000");
+        System.setProperty(PipelinedMergeSortTask.OAK_INDEXER_PIPELINED_EAGER_MERGE_MAX_SIZE_TO_MERGE_MB, "1");
+        System.setProperty(PipelinedMergeSortTask.OAK_INDEXER_PIPELINED_EAGER_MERGE_MIN_FILES_TO_MERGE, "1000");
+
+        // Generate FFS
+        List<String> ffs = generateFFS(LINES_IN_FFS);
+        // Shuffle entries to simulate retrieving from MongoDB by an arbitrary order
+        Collections.shuffle(ffs);
+
+        // Generate the expected results by sorting using the node state entries comparator,
+        List<NodeStateHolder> nodesOrdered = sortAsNodeStateEntries(ffs);
+        // Convert back to a list of Strings
+        String[] expectedFFS = nodesOrdered.stream().map(f -> new String(f.getLine())).toArray(String[]::new);
```

Comment:
This uses the platforms default charset; should it use FLATFILESTORE_CHARSET, always UTF-8, or a configurable charset?

Reply:
Done.